pkg/cmd/cli/cmd/login.go

```diff
@@ -0,0 +1,248 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	kclientcmd "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/cli/config"
+	"github.com/openshift/origin/pkg/cmd/util"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+)
+
+type loginOptions struct {
+	username string
+	password string
+	server   string
+	project  string
+	context  string
+}
+
+func NewCmdLogin(f *osclientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &loginOptions{}
+
+	cmds := &cobra.Command{
+		Use:   "login [--username=<username>] [--password=<password>] [--server=<server>]",
+		Short: "Logs in and returns a session token",
+		Long: `Logs in to the OpenShift server and prints out a session token.
+
+Username and password can be provided through flags, the command will
+prompt for user input if not provided.
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			// fetch config and if a server were not provided, prompt for it
+			clientCfg, err := f.OpenShiftClientConfig.ClientConfig()
+			if isServerNotFound(err) {
+				defaultServer := osclientcmd.DefaultClusterURL
+				promptedServer := util.PromptForStringWithDefault(os.Stdin, defaultServer, "OpenShift is not configured. Please provide the server URL or hit <enter> to use '%v': ", defaultServer)
+				serverFlag := cmd.Flags().Lookup("server")
+				if serverFlag != nil {
+					serverFlag.Value.Set(promptedServer)
+					if err := cmd.Execute(); err != nil {
+						os.Exit(1)
+					} else {
+						os.Exit(0)
+					}
+				}
+			}
+			checkErr(err)
+
+			// try to find a config file from default locations, or create a new one if not found
+			configStore, err := config.GetConfigFromDefaultLocations(clientCfg, cmd)
+			checkErr(err)
+
+			// validate flags
+			checkErr(options.validate(configStore, clientCfg))
+
+			// Check to see if we're already signed in (with the user provided by --username)
+			// If so, simply make sure that .kubeconfig has that information
+			if userFullName, err := whoami(clientCfg); err == nil && (!options.usernameProvided() || (options.usernameProvided() && options.username == userFullName)) {
```

Comment:
Why would fullName == `--username`?  Seems like you'd want to know if `--username` == `strings.Split(user.Name, ":")[1:]`